compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+    /// pass.
+    assignment_order: Vec<Local>,
+}
+
+impl SsaLocals {
+    fn new(body: &Body<'_>) -> SsaLocals {
+        let direct_uses = IndexVec::from_elem(0, &body.local_decls);
+        let assignment_order = Vec::new();
+
+        let assignments = IndexVec::from_elem(Set1::Empty, &body.local_decls);
+        let mut this = SsaLocals { assignments, direct_uses, assignment_order };
+
+        for (local, decl) in body.local_decls.iter_enumerated() {
+            if matches!(body.local_kind(local), LocalKind::Arg) {
+                this.assignments[local] = Set1::One(LocationExtended::Arg);
+            }
+            // Do not bother tracking non-pointers.
+            if !decl.ty.is_any_ptr() {
+                this.assignments[local] = Set1::Many;
+            }
+        }
+
+        for (bb, data) in traversal::reverse_postorder(body) {
+            this.visit_basic_block_data(bb, data);
+        }
+
+        for var_debug_info in &body.var_debug_info {
+            this.visit_var_debug_info(var_debug_info);
+        }
+
+        debug!(?this.assignments);
+        debug!(?this.direct_uses);
+
+        this.assignment_order.retain(|&local| matches!(this.assignments[local], Set1::One(_)));
+        debug!(?this.assignment_order);
+
+        this
+    }
+
+    /// Compute the equivalence classes for locals, based on copy statements.
+    ///
+    /// The returned vector maps each local to the one it copies. In the following case:
+    ///   _a = &mut _0
+    ///   _b = _a
+    ///   _c = _a
+    ///   _d = _c
+    /// We return the mapping
+    ///   _a => _a // not a copy so, represented by itself
+    ///   _b => _a
+    ///   _c => _a
+    ///   _d => _a // transitively through _c
+    fn compute_copy_classes(&mut self, body: &Body<'_>) -> IndexVec<Local, Local> {
+        let mut copies = IndexVec::from_fn_n(|l| l, body.local_decls.len());
+
+        for &local in &self.assignment_order {
+            debug!(?local);
+
+            // This is not SSA: mark that we don't know the value.
+            debug!(assignments = ?self.assignments[local]);
+            let Set1::One(LocationExtended::Plain(loc)) = self.assignments[local] else {
+                continue;
+            };
+
+            // `loc` must point to a direct assignment to `local`.
+            let Either::Left(stmt) = body.stmt_at(loc) else { bug!() };
+            let Some((_target, rvalue)) = stmt.kind.as_assign() else { bug!() };
+            assert_eq!(_target.as_local(), Some(local));
+
+            // This is a copy, just use the value we have in store for the previous one.
+            // As we are visiting in `assignment_order`, ie. reverse postorder, `rhs` should
+            // have been visited before.
+            if let Rvalue::Use(Operand::Copy(place) | Operand::Move(place))
+            | Rvalue::CopyForDeref(place) = rvalue
+                && let Some(rhs) = place.as_local()
```

Comment:
this formatting is quite odd. Maybe continue the pattern used above?

```suggestion
            let Rvalue::Use(Operand::Copy(place) | Operand::Move(place)) | Rvalue::CopyForDeref(place) = rvalue else {
                continue;
            };
            let Some(rhs) = place.as_local() else {
                continue;
            };
```